atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/ring/RingGraph.java

```diff
@@ -0,0 +1,120 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import one.util.streamex.EntryStream;
+
+public final class RingGraph {
+    private final Map<Integer, Integer> ring;
+
+    private RingGraph(Map<Integer, Integer> ring) {
+        this.ring = ring;
+    }
+
+    public static RingGraph from(Map<Integer, Integer> ring) throws RingValidationException {
+        validateGraphIsRing(ring);
+        return new RingGraph(ring);
+    }
+
+    public static RingGraph fromPartial(Map<Integer, Optional<Integer>> ring) throws RingValidationException {
+        if (isEmpty(ring)) {
+            return RingGraph.generateNewRing(ring.keySet());
+        }
+
+        Preconditions.checkArgument(
+                !anyEmpty(ring),
+                "Graph contains missing entries, thus cannot be made into a ring.",
+                SafeArg.of("ring", ring));
+
+        return from(EntryStream.of(ring).mapValues(Optional::get).toMap());
+    }
+
+    public static RingGraph create(int size) {
+        return generateNewRing(IntStream.range(0, size).boxed().collect(Collectors.toSet()));
+    }
+
+    /**
+     * Checks that the ring is correct by verifying:
+     * (1) There are no missing entries
+     * (2) There are no sub-cycles
+     * (3) The ring is connected
+     *
+     * This is simply done by iterating over the length of the ring, and validating that we visited each node.
+     * To ensure that the ring is connected, we do not track our entry into the ring (i.e. the first node we visit),
+     * as we expect to have it be visited again as the last node.
+     */
+    public static void validateGraphIsRing(Map<Integer, Integer> ring) throws RingValidationException {
+        Integer initialNode = ring.keySet().iterator().next();
+        Set<Integer> remainingNodes = new HashSet<>(ring.keySet());
+        int maxIterations = ring.keySet().size();
+        Integer nextNode = ring.get(initialNode);
+        for (int iterations = 0; iterations < maxIterations; iterations++) {
+            // If we reference a node that does not exist in our ring, it means we are missing data
+            if (!ring.containsKey(nextNode)) {
+                RingValidationException.throwMissingEntries(ring);
+            }
+
+            // If the node already has been visited, it means we're in a cycle within our ring
+            if (!remainingNodes.remove(nextNode)) {
+                RingValidationException.throwCycle(ring);
+            }
+
+            nextNode = ring.get(nextNode);
+        }
+    }
+
+    public RingGraph generateNewRing() {
+        return generateNewRing(ring.keySet());
+    }
+
+    /**
+     * Returns a valid new ring with the same nodes but randomly generated edges.
+     */
+    private static RingGraph generateNewRing(Set<Integer> nodes) {
```

Comment:
nice use of cycle! wished there was an API to fluently zip a stream and a lagged/forward version of it